dotnet/src/Connectors/Connectors.AI.Oobabooga/TextCompletion/OoobaboogaTextCompletion.cs

```diff
@@ -0,0 +1,232 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.WebSockets;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.Oobabooga.TextCompletion;
+
+/// <summary>
+/// Oobabooga text completion service API.
+/// Adapted from https://github.com/oobabooga/text-generation-webui/tree/main/api-examples
+/// </summary>
+public sealed class OobaboogaTextCompletion : ITextCompletion, IDisposable
+{
+    private const string HttpUserAgent = "Microsoft-Semantic-Kernel";
+    private const string BlockingUriPath = "/api/v1/generate";
+    private const string StreamingUriPath = "/api/v1/stream";
+
+    private readonly Uri _endpoint;
+    private readonly int _blockingPort;
+    private readonly int _streamingPort;
+    private readonly HttpClient _httpClient;
+    private readonly ClientWebSocket _webSocket;
+    private readonly HttpClientHandler? _httpClientHandler;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="OobaboogaTextCompletion"/> class.
+    /// </summary>
+    /// <param name="endpoint">Endpoint for service API call.</param>
+    /// <param name="blockingPort">The port for blocking requests.</param>
+    /// <param name="streamingPort">The port for streaming requests.</param>
+    /// <param name="httpClientHandler">Instance of <see cref="HttpClientHandler"/> to setup specific scenarios.</param>
+    public OobaboogaTextCompletion(Uri endpoint, int blockingPort, int streamingPort, HttpClientHandler httpClientHandler)
+    {
+        Verify.NotNull(endpoint);
+
+        this._endpoint = endpoint;
+        this._blockingPort = blockingPort;
+        this._streamingPort = streamingPort;
+
+        this._httpClient = new(httpClientHandler);
+        this._httpClient.DefaultRequestHeaders.Add("User-Agent", HttpUserAgent);
+        this._webSocket = new ClientWebSocket();
```

Comment:
See #1357. 

Note that websockets work differently, so User-Agent will be permanently overriden here unlike the blocking solution, which only affects the local request. I guess this isn't much of a an issue, since websocketclient shouldn't really be shared with other usages, and the ctor parameter simply allows to add more options to it, but it still feels a bit inconfortable to affect the input client.